File: elixir/twelve-days/lib/twelve_days.ex
defmodule TwelveDays do
  @presents_per_day [
    "a Partridge in a Pear Tree",
    "two Turtle Doves",
    "three French Hens",
    "four Calling Birds",
    "five Gold Rings",
    "six Geese-a-Laying",
    "seven Swans-a-Swimming",
    "eight Maids-a-Milking",
    "nine Ladies Dancing",
    "ten Lords-a-Leaping",
    "eleven Pipers Piping",
    "twelve Drummers Drumming"
  ]

  @order [
    "first",
    "second",
    "third",
    "fourth",
    "fifth",
    "sixth",
    "seventh",
    "eighth",
    "ninth",
    "tenth",
    "eleventh",
    "twelfth"
  ]

  @doc """
  Given a `number`, return the song's verse for that specific day, including
  all gifts for previous days in the same line.
  """
  @spec verse(number :: integer) :: String.t()
  def verse(number) do
    "#{on_day(number)} #{gifts_for(number)}."
  end

  @doc """
  Given a `starting_verse` and an `ending_verse`, return the verses for each
  included day, one per line.
  """
  @spec verses(starting_verse :: integer, ending_verse :: integer) :: String.t()
  def verses(starting_verse, ending_verse) do
    starting_verse..ending_verse
    |> Enum.map(&verse/1)
    |> Enum.join("\n")
  end

  @doc """
  Sing all 12 verses, in order, one verse per line.
  """
  @spec sing() :: String.t()
  def sing,
    do: verses(1, 12)

  @spec gift_for(non_neg_integer()) :: String.t()
  defp gift_for(day_number),
    do: @presents_per_day |> Enum.at(day_number - 1)

  defp gifts_for(1),
    do: gift_for(1)

  defp gifts_for(day_number) do
    previous_gifts = day_number..2 |> Enum.map(&gift_for/1) |> Enum.join(", ")
    "#{previous_gifts}, and #{gift_for(1)}"
  end

  @spec ordenal_for(non_neg_integer()) :: String.t()
  defp ordenal_for(number),
    do: @order |> Enum.at(number - 1)

  @spec on_day(non_neg_integer()) :: String.t()
  defp on_day(number),
    do: "On the #{ordenal_for(number)} day of Christmas my true love gave to me:"
end

File: lua/twelve-days/twelve-days.lua
--- A module that provides the lyrics for "The Twelve Days of Christmas".
local twelve_days <const> = {}

--- A constant table containing all twelve verses of the song.
local VERSES <const> = {
    'On the first day of Christmas my true love gave to me: a Partridge in a Pear Tree.',
    'On the second day of Christmas my true love gave to me: two Turtle Doves, and a Partridge in a Pear Tree.',
    'On the third day of Christmas my true love gave to me: three French Hens, two Turtle Doves, and a Partridge in a Pear Tree.',
    'On the fourth day of Christmas my true love gave to me: four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree.',
    'On the fifth day of Christmas my true love gave to me: five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree.',
    'On the sixth day of Christmas my true love gave to me: six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree.',
    'On the seventh day of Christmas my true love gave to me: seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree.',
    'On the eighth day of Christmas my true love gave to me: eight Maids-a-Milking, seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree.',
    'On the ninth day of Christmas my true love gave to me: nine Ladies Dancing, eight Maids-a-Milking, seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree.',
    'On the tenth day of Christmas my true love gave to me: ten Lords-a-Leaping, nine Ladies Dancing, eight Maids-a-Milking, seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree.',
    'On the eleventh day of Christmas my true love gave to me: eleven Pipers Piping, ten Lords-a-Leaping, nine Ladies Dancing, eight Maids-a-Milking, seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree.',
    'On the twelfth day of Christmas my true love gave to me: twelve Drummers Drumming, eleven Pipers Piping, ten Lords-a-Leaping, nine Ladies Dancing, eight Maids-a-Milking, seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree.'
}

--- Retrieves a range of verses from the song.
--- The verses are returned as a table of strings.
--- @param start_verse integer The 1-based index of the first verse.
--- @param end_verse integer The 1-based index of the last verse.
--- @return string[] A table containing the requested verses.
function twelve_days.recite(start_verse, end_verse)
    return { table.unpack(VERSES, start_verse, end_verse) }
end

return twelve_days